File: 2413-smallest-number-in-infinite-set/2413-smallest-number-in-infinite-set.py
class SmallestInfiniteSet:
    def __init__(self):
       self.curr_min=1
       self.heap=[] 

    def popSmallest(self) -> int:
        if self.heap:
            return heapq.heappop(self.heap)
        self.curr_min+=1
        return self.curr_min-1
        

    def addBack(self, num: int) -> None:
        if self.curr_min>num and num not in set(self.heap):
            heapq.heappush(self.heap,num)

        


# Your SmallestInfiniteSet object will be instantiated and called as such:
# obj = SmallestInfiniteSet()
# param_1 = obj.popSmallest()
# obj.addBack(num)

File: 2413-smallest-number-in-infinite-set/2413-smallest-number-in-infinite-set.java
class SmallestInfiniteSet {
    private PriorityQueue<Integer> heap;
    private int min;
    public SmallestInfiniteSet() {
        heap=new PriorityQueue<>();
        min=1;
    }
    public int popSmallest() {
        if(!heap.isEmpty()){
            return heap.poll();
        }   
        min++;
        return min-1;
    }
    
    public void addBack(int num) {
        if(min>num && !heap.contains(num)){
            heap.add(num);
        }
    }
}

/**
 * Your SmallestInfiniteSet object will be instantiated and called as such:
 * SmallestInfiniteSet obj = new SmallestInfiniteSet();
 * int param_1 = obj.popSmallest();
 * obj.addBack(num);
 */